libs/parallelism/algorithms/include/hpx/parallel/algorithms/unique.hpp

```diff
@@ -213,147 +682,101 @@ namespace hpx { namespace parallel { inline namespace v1 {
-        Proj&& proj = Proj())
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_iterator_v<FwdIter> &&
+            traits::is_projected<Proj, FwdIter>::value &&
+            traits::is_indirect_callable<ExPolicy, Pred,
+                    traits::projected<Proj, FwdIter>,
+                    traits::projected<Proj, FwdIter>>::value
+        )>
+    // clang-format on
+    HPX_DEPRECATED_V(1, 8,
+        "hpx::parallel::unique is deprecated, use "
+        "hpx::unique instead")
+        typename util::detail::algorithm_result<ExPolicy, FwdIter>::type
+        unique(ExPolicy&& policy, FwdIter first, FwdIter last,
+            Pred&& pred = Pred(), Proj&& proj = Proj())
     {
-        static_assert((hpx::traits::is_forward_iterator<FwdIter>::value),
+        static_assert((hpx::traits::is_forward_iterator_v<FwdIter>),
             "Required at least forward iterator.");
 
+#if defined(HPX_GCC_VERSION) && HPX_GCC_VERSION >= 100000
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
         return detail::unique<FwdIter>().call(std::forward<ExPolicy>(policy),
             first, last, std::forward<Pred>(pred), std::forward<Proj>(proj));
+#if defined(HPX_GCC_VERSION) && HPX_GCC_VERSION >= 100000
+#pragma GCC diagnostic pop
+#endif
     }
 
+    template <typename I, typename O>
+    using unique_copy_result = util::in_out_result<I, O>;
+
     /////////////////////////////////////////////////////////////////////////////
     // unique_copy
     namespace detail {
         /// \cond NOINTERNAL
 
         // sequential unique_copy with projection function
-        template <typename FwdIter, typename OutIter, typename Pred,
-            typename Proj>
-        std::pair<FwdIter, OutIter> sequential_unique_copy(FwdIter first,
-            FwdIter last, OutIter dest, Pred&& pred, Proj&& proj,
+        template <typename FwdIter, typename Sent, typename OutIter,
+            typename Pred, typename Proj>
+        unique_copy_result<FwdIter, OutIter> sequential_unique_copy(
+            FwdIter first, Sent last, OutIter dest, Pred&& pred, Proj&& proj,
             std::true_type)
         {
             if (first == last)
-                return std::make_pair(std::move(last), std::move(dest));
+                return unique_copy_result<FwdIter, OutIter>{
+                    std::move(first), std::move(dest)};
 
             FwdIter base = first;
 
             *dest++ = *first;
 
             while (++first != last)
             {
-                if (!hpx::util::invoke(pred, hpx::util::invoke(proj, *base),
-                        hpx::util::invoke(proj, *first)))
+                if (!HPX_INVOKE(pred, HPX_INVOKE(proj, *base),
+                        HPX_INVOKE(proj, *first)))
                 {
                     base = first;
                     *dest++ = *first;
                 }
             }
```

Comment:
More optimizations may be possible by avoiding to invoke the projection on `base` (which is a copy of `first` for which the projection was already invoked).